pca_cca.go

```diff
@@ -0,0 +1,239 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import (
+	"errors"
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix"
+	"github.com/gonum/matrix/mat64"
+)
+
+// PrincipalComponents returns the principal component direction vectors and
+// the column variances of the principal component scores, vecs * a, computed
+// using the singular value decomposition of the input. The input a is an n×d
+// matrix where each row is an observation and each column represents a variable.
+//
+// PrincipalComponents centers the variables but does not scale the variance.
+//
+// The slice weights is used to weight the observations. If weights is nil,
+// each weight is considered to have a value of one, otherwise the length of
+// weights must match the number of observations or PrincipalComponents will
+// panic.
+//
+// On successful completion, the principal component direction vectors are
+// returned in vecs as a d×min(n, d) matrix, and the variances are returned in
+// vars as a min(n, d)-long slice in descending sort order.
+//
+// If no singular value decomposition is possible, vecs and vars are returned
+// nil and ok is returned false.
+func PrincipalComponents(a mat64.Matrix, weights []float64) (vecs *mat64.Dense, vars []float64, ok bool) {
+	n, d := a.Dims()
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	svd, ok := svdFactorizeCentered(a, weights)
+	if !ok {
+		return nil, nil, false
+	}
+
+	vecs = &mat64.Dense{}
+	vecs.VFromSVD(svd)
+	if n < d {
+		// Don't retain columns that are not valid direction vectors.
+		vecs.Clone(vecs.View(0, 0, d, n))
+	}
+	vars = svd.Values(nil)
+	var f float64
+	if weights == nil {
+		f = 1 / float64(n-1)
+	} else {
+		f = 1 / (floats.Sum(weights) - 1)
+	}
+	for i, v := range vars {
+		vars[i] = f * v * v
+	}
+	return vecs, vars, true
+}
+
+// scaleColsReciSqrt is a helper for scaling the columns of cols
+// by the reciprocal square-root of vals
+func scaleColsReciSqrt(cols *mat64.Dense, vals []float64) {
+	if cols == nil {
+		panic("stat: input nil")
+	}
+	n, d := cols.Dims()
+	if len(vals) != d {
+		panic("stat: input length mismatch")
+	}
+	col := make([]float64, n)
+	for j := 0; j < d; j++ {
+		mat64.Col(col, j, cols)
+		floats.Scale(math.Sqrt(1/vals[j]), col)
+		cols.SetCol(j, col)
+	}
+}
+
+// CC holds the relevant SVDs for canonical correlation analysis
+type CC struct {
+	n int
```

Comment:
but `x`, `y`, and `c` don't? None of these are outwards facing. 
`n` is the number of observations in the data, so the number of 
rows in x and/or y. `nObs` maybe?